src/combineReducers.js

```diff
@@ -26,7 +25,7 @@ function getUnexpectedStateShapeWarningMessage(inputState, reducers, action, une
     )
   }
 
-  if (!isPlainObject(inputState)) {
+  if (!inputState || inputState.constructor !== Object) {
```

Comment:
I don't think this would work across iframes or multiple realms. Does it? Are we sure that this is a superset of Lodash's `true` answer rather than a subset or intersection?

Reply:
> I don't think this would work across iframes or multiple realms. Does it?

No it wouldn't. Or objects created with `Object.create(null)`

Reply:
Yes, I know about this. Hence the PR instead of just blasting this into master and calling it a day 😄 

It's not a use case I'm familiar with, so I'd love some help understanding it and coming up with some examples or even tests for it.

Reply:
Sry for the noob question but why `inputState.constructor !== Object` instead of `typeof inputState === 'object'`

Reply:
@phsantiago Because object type can represent any object `new Number` or `new Array` for example. Instance of `Object` is always plain.

Reply:
I'm very newbie. Would you explain what is the down side if the action is not plain?
Thanks in advance.

Reply:
@aminpaks It means it's not serializable -> may have side effects -> not predictable. And maybe redux devtools won't work too :)

Reply:
I'm very new to Redux, I actually started a few months back using it. It's an amazing approach and thanks for all your work. We are using heavily Typescript to statically type check in our codebase, and at some point I realized that the boilerplate in defining the actions and their types is killing our team (maybe everyone!).

This is an example of defining [actions in NgRx](https://github.com/ngrx/example-app/blob/master/src/app/actions/collection.ts) project. Literally they are defined in three different place for different purposes. Looking at this recently I took [another approach](https://github.com/aminpaks/angular-redux-router/blob/master/src/app/store/store.utils.ts) to keep track of actions and its typings:
```ts
// Already defined in store definitions
import { ItemX } from '...';

// This action will have a payload as type of `ItemX`
export const AHigherOrderAction = actionCreator<ItemX>('[Project X] A Higher Order Action');


... 
// In reducer
import { AHigherOrderAction } from './actions';
...
export function reducer(state: ATypeForThisPortionOfStore = initialState, action: Action) {
  switch(action.type) {
  case AHigherOrderAction.type:
    return { ...state }
  default:
    return state;
}


...
// dispatch the action
store.dispatch(new AHigherOrderAction(aVariableWithAProperType));
...
```
Currently to be able to make it work I'm converting these actions (functions) to plain objects by a middleware.
I understand there is a performance concern in converting part (I did [this benchmark](https://jsperf.com/convert-higher-order-actions-to-plain-object-actions) to see what's the best approach), but I believe it's worth using this approach in terms the amount of code we save.

Long story short, you already mentioned some reasons that the actions are supposed to be plain. What if redux api can offer something to convert these actions internally. Cause currently we have some issues in which other middlewares are skipping the conversion and this crashes Redux.

Any thoughts?

Reply:
@aminpaks I already passed through this overengineering with actions as classes. Currently my boilerplate is
```ts
// actions.js
export type Action
  = { type: 'DO_ONE_THING' }
  | { type: 'DO_ANOTHER_THING' };

// reducers.js
import { Action } from './actions.js';

export const reducer = (state: State, action: Action) => {
  switch () {
    case 'DO_ONE_THING':
      return state;
    default:
      return state;
  }
};
```

Reply:
There is not any benefits in classes.

Reply:
@TrySound in you example in the reducer file => Can IDE with intellisence resolve DO_ONE_THING ?

Reply:
@nasreddineskandrani Nope, but type checker will say what's wrong. For IDE fans there is small boilerplate like this
```js
export const DO_ONE_THING = 'DO_ONE_THING';
```

Reply:
@TrySound Correct me if I'm wrong, according to you:
> It means it's not serializable -> may have side effects -> not predictable. And maybe redux devtools won't work too :)

If this won't break anything developer should be able to bypass it, no?

Reply:
@TrySound 
please read enum string introduced in ts2.4 section here: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-4.html
enum string => give you intellisense ability (save time) and type checking.

Reply:
@aminpaks , @nasreddineskandrani , @TrySound : I do appreciate the enthusiasm and interest, but this discussion would probably be better off in a different location.  

The Reactiflux chat channels at https://www.reactiflux.com always have plenty of people around who are happy to discuss Redux-related topics.  There's somewhat less discussion of TypeScript or NgRX, but there might be some people who can discuss those.

Reply:
@markerikson I hesitate to comment but since the code was in a sub thread. I gived my 2 cents. Thanks for the link I ll be happy to exchange about redux their.